node/service/src/lib.rs

```diff
@@ -428,28 +513,78 @@ pub fn new_full<RuntimeApi, Executor>(
 		})
 		.collect();
 
-	let (overseer, overseer_handler) = real_overseer(leaves, prometheus_registry.as_ref(), spawner)?;
-	let overseer_handler_clone = overseer_handler.clone();
+	let authority_discovery_service = if role.is_authority() || is_collator {
+		use sc_network::Event;
+		use futures::StreamExt;
 
-	task_manager.spawn_essential_handle().spawn_blocking("overseer", Box::pin(async move {
-		use futures::{pin_mut, select, FutureExt};
+		if authority_discovery_disabled {
+			Err("Authority discovery is mandatory for a validator.")?;
+		}
 
-		let forward = polkadot_overseer::forward_events(overseer_client, overseer_handler_clone);
+		let authority_discovery_role = if role.is_authority() {
+			authority_discovery::Role::PublishAndDiscover(
+				keystore_container.keystore(),
+			)
+		} else {
+			// don't publish our addresses when we're only a collator
+			authority_discovery::Role::Discover
+		};
+		let dht_event_stream = network.event_stream("authority-discovery")
+			.filter_map(|e| async move { match e {
+				Event::Dht(e) => Some(e),
+				_ => None,
+			}});
+		let (worker, service) = authority_discovery::new_worker_and_service(
+			client.clone(),
+			network.clone(),
+			Box::pin(dht_event_stream),
+			authority_discovery_role,
+			prometheus_registry.clone(),
+		);
 
-		let forward = forward.fuse();
-		let overseer_fut = overseer.run().fuse();
+		task_manager.spawn_handle().spawn("authority-discovery-worker", worker.run());
+		Some(service)
+	} else {
+		None
+	};
 
-		pin_mut!(overseer_fut);
-		pin_mut!(forward);
+	// we'd say let overseer_handler = authority_discovery_service.map(|authority_discovery_service|, ...),
+	// but in that case we couldn't use ? to propagate errors
+	let overseer_handler = if let Some(authority_discovery_service) = authority_discovery_service {
+		let (overseer, overseer_handler) = real_overseer(
+			leaves,
+			keystore_container.sync_keystore(),
+			overseer_client.clone(),
+			availability_config?,
+			network.clone(),
+			authority_discovery_service,
+			prometheus_registry.as_ref(),
+			spawner,
+		)?;
+		let overseer_handler_clone = overseer_handler.clone();
+
+		task_manager.spawn_essential_handle().spawn_blocking("overseer", Box::pin(async move {
+			use futures::{pin_mut, select, FutureExt};
+
+			let forward = polkadot_overseer::forward_events(overseer_client, overseer_handler_clone);
 
-		loop {
-			select! {
-				_ = forward => break,
-				_ = overseer_fut => break,
-				complete => break,
+			let forward = forward.fuse();
+			let overseer_fut = overseer.run().fuse();
+
+			pin_mut!(overseer_fut);
+			pin_mut!(forward);
```

Comment:
```suggestion
			let mut forward = forward.fuse();
			let mut overseer_fut = overseer.run().fuse();
```

Should do the same?

Reply:
Nope:

```
error[E0277]: `from_generator::GenFuture<[static generator@forward_events::{closure#0} for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6, 't7, 't8, 't9, 't10> {ResumeTy, Arc<sc_service::client::Client<sc_client_db::Backend<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>, LocalCallExecutor<sc_client_db::Backend<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>, NativeExecutor<Executor>>, sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>, RuntimeApi>>, OverseerHandler, sp_utils::mpsc::inner::TracingUnboundedReceiver<FinalityNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>, sp_utils::mpsc::inner::TracingUnboundedReceiver<BlockImportNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>, Next<'r, sp_utils::mpsc::inner::TracingUnboundedReceiver<FinalityNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>>, Next<'s, sp_utils::mpsc::inner::TracingUnboundedReceiver<BlockImportNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>>, [closure@forward_events::{closure#0}::{closure#0}], futures::future::PollFn<[closure@forward_events::{closure#0}::{closure#0}]>, (), forward_events::{closure#0}::__PrivResult<std::option::Option<FinalityNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>, std::option::Option<BlockImportNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>>, std::option::Option<FinalityNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>, FinalityNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>, &'t8 mut OverseerHandler, BlockInfo, impl futures::Future, std::option::Option<BlockImportNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>, BlockImportNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>, impl futures::Future}]>` cannot be unpinned
   --> node/service/src/lib.rs:591:5
    |
591 |                   select! {
    |  _________________^
592 | |                     _ = forward => break,
593 | |                     _ = overseer_fut => break,
594 | |                     complete => break,
595 | |                 }
    | |_________________^ within `futures_util::future::future::fuse::_::__Fuse<'_, impl futures::Future>`, the trait `Unpin` is not implemented for `from_generator::GenFuture<[static generator@forward_events::{closure#0} for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6, 't7, 't8, 't9, 't10> {ResumeTy, Arc<sc_service::client::Client<sc_client_db::Backend<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>, LocalCallExecutor<sc_client_db::Backend<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>, NativeExecutor<Executor>>, sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>, RuntimeApi>>, OverseerHandler, sp_utils::mpsc::inner::TracingUnboundedReceiver<FinalityNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>, sp_utils::mpsc::inner::TracingUnboundedReceiver<BlockImportNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>, Next<'r, sp_utils::mpsc::inner::TracingUnboundedReceiver<FinalityNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>>, Next<'s, sp_utils::mpsc::inner::TracingUnboundedReceiver<BlockImportNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>>, [closure@forward_events::{closure#0}::{closure#0}], futures::future::PollFn<[closure@forward_events::{closure#0}::{closure#0}]>, (), forward_events::{closure#0}::__PrivResult<std::option::Option<FinalityNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>, std::option::Option<BlockImportNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>>, std::option::Option<FinalityNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>, FinalityNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>, &'t8 mut OverseerHandler, BlockInfo, impl futures::Future, std::option::Option<BlockImportNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>, BlockImportNotification<sp_runtime::generic::Block<sp_runtime::generic::Header<u32, BlakeTwo256>, OpaqueExtrinsic>>, impl futures::Future}]>`
    | 
   ::: /home/coriolinus/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.5/src/async_await/mod.rs:43:24
    |
43  |   pub fn assert_unpin<T: Unpin>(_: &T) {}
    |                          ----- required by this bound in `futures_util::async_await::assert_unpin`
    |
    = note: required because it appears within the type `impl futures::Future`
    = note: required because it appears within the type `impl futures::Future`
    = note: required because it appears within the type `std::option::Option<impl futures::Future>`
    = note: required because it appears within the type `futures_util::future::future::fuse::_::__Fuse<'_, impl futures::Future>`
    = note: required because of the requirements on the impl of `Unpin` for `futures::future::Fuse<impl futures::Future>`
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0277]: `from_generator::GenFuture<[static generator@Overseer::<S>::run::{closure#0} for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6, 't7, 't8, 't9, 't10, 't11, 't12, 't13, 't14, 't15, 't16, 't17, 't18, 't19, 't20, 't21> {ResumeTy, Overseer<SpawnTaskHandle>, Vec<(H256, u32)>, ActiveLeavesUpdate, &'r mut Overseer<SpawnTaskHandle>, OverseerSignal, impl futures::Future, (), &'t0 mut futures::channel::mpsc::Receiver<polkadot_overseer::Event>, futures::channel::mpsc::Receiver<polkadot_overseer::Event>, Next<'t1, futures::channel::mpsc::Receiver<polkadot_overseer::Event>>, &'t2 mut Next<'t3, futures::channel::mpsc::Receiver<polkadot_overseer::Event>>, futures_util::async_await::poll::PollOnce<&'t4 mut Next<'t5, futures::channel::mpsc::Receiver<polkadot_overseer::Event>>>, Poll<std::option::Option<polkadot_overseer::Event>>, polkadot_overseer::Event, polkadot_node_subsystem::messages::AllMessages, impl futures::Future, impl futures::Future, BlockInfo, impl futures::Future, impl futures::Future, &'t9 mut streamunordered::StreamUnordered<futures::channel::mpsc::Receiver<polkadot_overseer::ToOverseer>>, streamunordered::StreamUnordered<futures::channel::mpsc::Receiver<polkadot_overseer::ToOverseer>>, Next<'t10, streamunordered::StreamUnordered<futures::channel::mpsc::Receiver<polkadot_overseer::ToOverseer>>>, &'t11 mut Next<'t12, streamunordered::StreamUnordered<futures::channel::mpsc::Receiver<polkadot_overseer::ToOverseer>>>, futures_util::async_await::poll::PollOnce<&'t13 mut Next<'t14, streamunordered::StreamUnordered<futures::channel::mpsc::Receiver<polkadot_overseer::ToOverseer>>>>, Poll<std::option::Option<(streamunordered::StreamYield<futures::channel::mpsc::Receiver<polkadot_overseer::ToOverseer>>, usize)>>, polkadot_overseer::ToOverseer, &'t15 mut FuturesUnordered<Pin<Box<(dyn futures::Future<Output = sc_service::Result<(), SubsystemError>> + std::marker::Send + 't16)>>>, FuturesUnordered<Pin<Box<(dyn futures::Future<Output = sc_service::Result<(), SubsystemError>> + std::marker::Send + 't17)>>>, Next<'t18, FuturesUnordered<Pin<Box<(dyn futures::Future<Output = sc_service::Result<(), SubsystemError>> + std::marker::Send + 't19)>>>>, futures_util::async_await::poll::PollOnce<Next<'t20, FuturesUnordered<Pin<Box<(dyn futures::Future<Output = sc_service::Result<(), SubsystemError>> + std::marker::Send + 't21)>>>>>, Poll<std::option::Option<sc_service::Result<(), SubsystemError>>>, sc_service::Result<(), SubsystemError>, futures_util::async_await::pending::PendingOnce}]>` cannot be unpinned
   --> node/service/src/lib.rs:591:5
    |
591 |                   select! {
    |  _________________^
592 | |                     _ = forward => break,
593 | |                     _ = overseer_fut => break,
594 | |                     complete => break,
595 | |                 }
    | |_________________^ within `futures_util::future::future::fuse::_::__Fuse<'_, impl futures::Future>`, the trait `Unpin` is not implemented for `from_generator::GenFuture<[static generator@Overseer::<S>::run::{closure#0} for<'r, 's, 't0, 't1, 't2, 't3, 't4, 't5, 't6, 't7, 't8, 't9, 't10, 't11, 't12, 't13, 't14, 't15, 't16, 't17, 't18, 't19, 't20, 't21> {ResumeTy, Overseer<SpawnTaskHandle>, Vec<(H256, u32)>, ActiveLeavesUpdate, &'r mut Overseer<SpawnTaskHandle>, OverseerSignal, impl futures::Future, (), &'t0 mut futures::channel::mpsc::Receiver<polkadot_overseer::Event>, futures::channel::mpsc::Receiver<polkadot_overseer::Event>, Next<'t1, futures::channel::mpsc::Receiver<polkadot_overseer::Event>>, &'t2 mut Next<'t3, futures::channel::mpsc::Receiver<polkadot_overseer::Event>>, futures_util::async_await::poll::PollOnce<&'t4 mut Next<'t5, futures::channel::mpsc::Receiver<polkadot_overseer::Event>>>, Poll<std::option::Option<polkadot_overseer::Event>>, polkadot_overseer::Event, polkadot_node_subsystem::messages::AllMessages, impl futures::Future, impl futures::Future, BlockInfo, impl futures::Future, impl futures::Future, &'t9 mut streamunordered::StreamUnordered<futures::channel::mpsc::Receiver<polkadot_overseer::ToOverseer>>, streamunordered::StreamUnordered<futures::channel::mpsc::Receiver<polkadot_overseer::ToOverseer>>, Next<'t10, streamunordered::StreamUnordered<futures::channel::mpsc::Receiver<polkadot_overseer::ToOverseer>>>, &'t11 mut Next<'t12, streamunordered::StreamUnordered<futures::channel::mpsc::Receiver<polkadot_overseer::ToOverseer>>>, futures_util::async_await::poll::PollOnce<&'t13 mut Next<'t14, streamunordered::StreamUnordered<futures::channel::mpsc::Receiver<polkadot_overseer::ToOverseer>>>>, Poll<std::option::Option<(streamunordered::StreamYield<futures::channel::mpsc::Receiver<polkadot_overseer::ToOverseer>>, usize)>>, polkadot_overseer::ToOverseer, &'t15 mut FuturesUnordered<Pin<Box<(dyn futures::Future<Output = sc_service::Result<(), SubsystemError>> + std::marker::Send + 't16)>>>, FuturesUnordered<Pin<Box<(dyn futures::Future<Output = sc_service::Result<(), SubsystemError>> + std::marker::Send + 't17)>>>, Next<'t18, FuturesUnordered<Pin<Box<(dyn futures::Future<Output = sc_service::Result<(), SubsystemError>> + std::marker::Send + 't19)>>>>, futures_util::async_await::poll::PollOnce<Next<'t20, FuturesUnordered<Pin<Box<(dyn futures::Future<Output = sc_service::Result<(), SubsystemError>> + std::marker::Send + 't21)>>>>>, Poll<std::option::Option<sc_service::Result<(), SubsystemError>>>, sc_service::Result<(), SubsystemError>, futures_util::async_await::pending::PendingOnce}]>`
    | 
   ::: /home/coriolinus/.cargo/registry/src/github.com-1ecc6299db9ec823/futures-util-0.3.5/src/async_await/mod.rs:43:24
    |
43  |   pub fn assert_unpin<T: Unpin>(_: &T) {}
    |                          ----- required by this bound in `futures_util::async_await::assert_unpin`
    |
    = note: required because it appears within the type `impl futures::Future`
    = note: required because it appears within the type `impl futures::Future`
    = note: required because it appears within the type `std::option::Option<impl futures::Future>`
    = note: required because it appears within the type `futures_util::future::future::fuse::_::__Fuse<'_, impl futures::Future>`
    = note: required because of the requirements on the impl of `Unpin` for `futures::future::Fuse<impl futures::Future>`
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)
```